### ISSUE TITLE
Task005 pretrained file. Inconsistent dice score in summary.csv and the result of nnUNet_determine_postprocessing

### ISSUE THREAD

**Issue by ProfessorHuang** [MIC-DKFZ/nnUNet#484]
Hello Fabian,
I have downloaded the file Task005_Prostate.zip in Zenodo. After running the validation process of 3d_fullres, the postprocessed  5-fold cross-validation mean dice is [0.6838, 0.7918]. However, the 3d_fullres result provided in Task005_Prostate/ensembles/Task005_Prostate/summary.csv is [0.6678, 0.841].
By the way, I also tested it in in Task001, Task002 and Task004, the dice score after validation and dice score in the csv file is consistent. However, in Task005, they are not. The dice score in csv file seem to be correct, maybe something wrong in validation process. 
Below is the code I run after copying the downloaded file to right place:
`nnUNet_train 3d_fullres nnUNetTrainerV2 5 0 -val`
`nnUNet_train 3d_fullres nnUNetTrainerV2 5 1 -val`
`nnUNet_train 3d_fullres nnUNetTrainerV2 5 2 -val`
`nnUNet_train 3d_fullres nnUNetTrainerV2 5 3 -val`
`nnUNet_train 3d_fullres nnUNetTrainerV2 5 4 -val`
`nnUNet_determine_postprocessing -m 3d_fullres -t 5 -tr nnUNetTrainerV2`

**Comment by ProfessorHuang** [comment]
I am trying to compare my work with nnUNet on the 5-fold validation result in MSD. But I'm confused about how to get the appropriate result of nnUNet. Currently, I try to compare with the result provided by pretrained model, but the inconsistency problem really puzzle me. Can you give me some advice? :-）

**Comment by FabianIsensee** [comment]
Hi, you are right something is off with those weights. I am not sure what it could be though. I will keep looking

**Comment by ProfessorHuang** [comment]
Thank you Fabian.

> Hi, you are right something is off with those weights. I am not sure what it could be though. I will keep looking

**Comment by ProfessorHuang** [comment]
Hello, Fabian.
I guess the calculation of dice score in some special cases may be wrong. I don't check the code , but the result in summary.json is suspicious.
In case **prostate_18** and **prostate_32**, the ground truth of them have no label 2. So, I think the dice of these two cases should be NaN rather than zero. When calculate the mean dice, these two case's dice in label 2 had better to be ignored. 
When the mean dice of label 2 in summary.json divide by 30 and multiply by 32, the result is 0.7918*32/30=0.8446, which is close to the result in csv file.
The above is just my personal guess, I will check the code further, looking forward to your reply.

**Comment by FabianIsensee** [comment]
Hi,
as far as I can tell there  is no error in the way we compute the Dice.
Let's consider the following scenarios in your example above (i.e. label2 is not present in the reference):
1) there is no predicted label2. In this case, all the terms of the Dice are 0, resulting in a 0/0 situation which we define as nan
2) there are false positive predictions. In this case the nominator (2x the intersection) is zero, but the denominator is not. 0/X is mathematically valid and it is always 0.
When aggregating the metrics we use np.nanmean, so the nan values are ignored. Metric computation has always been this way in nnU-Net. I think this is the best way to do it, but it's certainly not perfect
Best,
Fabian

**Comment by FabianIsensee** [comment]
Whatever went wrong, I have now rerun the training from scratch. The Dice scores are:
        "1": 0.6681865172021348,
        "2": 0.843497121033962

I think that's close enough. I will update the weights in zenodo later today.
Best,
Fabian

**Comment by ProfessorHuang** [comment]
Thank you for your patient reply. It really helps.

**Comment by FabianIsensee** [comment]
weights are now updated. If you pull from master nnUNet_download... should use the correct link. You can also download them from zenodo: https://zenodo.org/record/4485926

**Comment by ProfessorHuang** [comment]
> Whatever went wrong, I have now rerun the training from scratch. The Dice scores are:
> "1": 0.6681865172021348,
> "2": 0.843497121033962
> 
> I think that's close enough. I will update the weights in zenodo later today.
> Best,
> Fabian

Sorry, Fabian. I'm not sure whether I did something wrong. I downloaded the latest file from zenodo and replace the file before. The dice score really changed but is close to the result before.
"1": 0.6858555860560707,
"2": 0.794438277370828
I did the same thing as before:
```
nnUNet_train 3d_fullres nnUNetTrainerV2 5 0 -val
nnUNet_train 3d_fullres nnUNetTrainerV2 5 1 -val
nnUNet_train 3d_fullres nnUNetTrainerV2 5 2 -val
nnUNet_train 3d_fullres nnUNetTrainerV2 5 3 -val
nnUNet_train 3d_fullres nnUNetTrainerV2 5 4 -val
nnUNet_determine_postprocessing -m 3d_fullres -t 5 -tr nnUNetTrainerV2
```
And then watch the mean dice score in RESULTS_FOLDER/nnUNet/3d_fullres/Task005_Prostate/nnUNetTrainerV2__nnUNetPlansv2.1/cv_niftis_postprocessed/summary.json

Best,
Ziyan

**Comment by FabianIsensee** [comment]
Please make sure to download from the correct DOI. The old one i s not replaced.
Use this link:
https://zenodo.org/record/4485926

**Comment by FabianIsensee** [comment]
wait - I am confused. I see the same issue lol

**Comment by FabianIsensee** [comment]
I found the error. It has nothing to do with nnU-Net. It's the data that is wrong!
cases 18 and 32 only have one label. The whole prostate consists of 'peripheral zone' (label 1). This is obviously incorrect - the annotators probably meant to label the whole prostate as 'central zone' (label 2) and didn't notice. I actually brought this to their attention back in 2018 and they changes that data. Somehow they seem to have changed it back to the incorrect labels. Weird stuff... I don't know why...

[prostate_18.nii.gz](https://github.com/MIC-DKFZ/nnUNet/files/5909408/prostate_18.nii.gz)
[prostate_32.nii.gz](https://github.com/MIC-DKFZ/nnUNet/files/5909409/prostate_32.nii.gz)

These are the correct gt masks (from my labelstr folder)

**Comment by ProfessorHuang** [comment]
Annotation is wrong,  that really suprised me!
I replaced the gt masks of cases 18 and 32, and then repeated the validation process. The dice socre is:
"1":  0.6700089854581439,
"2": 0.8434984599699235
That seems to be right.
Thank you for your time and selfless sharing！
best
Ziyan

**Comment by FabianIsensee** [comment]
I wouldn't be surprised ;-) There is a surprising amount of labeling errors in many datasets that even non-experts like me can spot. Annotation quality is very bad in the medical domain. There are only very few datasets with accurate annotations
